src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java

```diff
@@ -82,6 +85,16 @@ public boolean execute(Set<InetAddressAndPort> peers, Function<InetAddressAndPor
         if (peers == null || this.timeoutNanos < 0)
             return true;
 
+        // Check if there are any nodes which we know are running a version prior to 4.0.
+        // We use this intead of Gossiper::hasMajorVersion3Nodes because in the absence of version information for a peer
+        // we still prefer to run the startup connectivity check.
+        if (gossiper.isUpgradingFromVersionLowerThan(CassandraVersion.CASSANDRA_4_0))
```

Comment:
I do not think that you need to propagate Gossiper to constructor. Just use `Gossiper.instance` here. Gossiper is not an interface or similar that you need to access it like that.

Reply:
The only reason I added it to the constructor is to let me mock it out in the `StartupClusterConnectivityCheckerTest`. Happy to remove, but that would make testing the component a bit trickier.

Reply:
@isaacreath  you might still mock it but there would be one additional parameter to `execute` method which would be a function taking `CassandraVersion` and returning `boolean`:  `(version) -> yourMock.isUpgradingFromVersionLowerThan(version)`

Reply:
@smiklosovic Updated the PR to have `execute` take this as a predicate.